sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3307,12 +3307,18 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       _.containsPattern(NATURAL_LIKE_JOIN), ruleId) {
       case j @ Join(left, right, UsingJoin(joinType, usingCols), _, hint)
           if left.resolved && right.resolved && j.duplicateResolved =>
-        commonNaturalJoinProcessing(left, right, joinType, usingCols, None, hint)
```

Comment:
https://github.com/apache/spark/blob/9fa9d4b93176dcdf5f1e3d7c883956dc3f554508/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3492-L3499

commonNaturalJoinProcessing resolves `Join` to `Project(Join)` and discard the plan id, make this change to hold the plan id, otherwise following case will fail due to can not find the subplan:

```
left = spark.createDataFrame([Row(a=1)])
right = spark.createDataFrame([Row(a=1)])
df = left.join(right, on="a", how="left_outer")
df.withColumn("b", udf(lambda x: "x")(df.a)).    <- can not resolve `df.a`
```